src/typing.py

```diff
@@ -983,12 +979,24 @@ def __new__(cls, name, bases, namespace,
                          ", ".join(str(g) for g in gvars)))
                 tvars = gvars
 
+        orig_bases = bases
+        if extra is not None and type(extra) is abc.ABCMeta and extra not in bases:
+            bases = (extra,) + bases
+        bases = tuple(_gorg(b) if isinstance(b, GenericMeta) else b for b in bases)
+
+        # remove bare Generic from bases if there are other generic bases
+        if any(isinstance(b, GenericMeta) and b is not Generic for b in bases):
+            bases = tuple(b for b in bases if b is not Generic)
+        self = super().__new__(cls, name, bases, namespace, _root=True)
+
         self.__parameters__ = tvars
         self.__args__ = args
         self.__origin__ = origin
         self.__extra__ = extra
         # Speed hack (https://github.com/python/typing/issues/196).
         self.__next_in_mro__ = _next_in_mro(self)
+        if origin is None:
```

Comment:
I changed this a bit, to make it more clear and convenient to use. Now ``__orig_bases__`` is always created on subclassing, and then copied on ``__getitem__``. So that one don't need to use ``_gorg`` to find original bases, they are accessible immediately. I added a comment in code as well.